rules_128tech/pylint_plugins/mock_patch_autospec.py

```diff
@@ -0,0 +1,69 @@
+"""
+A plugin to detect and disallow the use of mock.patch or mocker.patch without
+setting autospec=True.
+
+Based on http://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html#write-a-checker
+"""
+
+import astroid
+import astroid.node_classes
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+
+
+def register(linter):
+    linter.register_checker(MockPatchAutospecChecker(linter))
+
+
+class MockPatchAutospecChecker(BaseChecker):
+    """A pylint checker to detect use of mock.patch without autospec=True"""
+
+    __implements__ = IAstroidChecker
+
+    name = "mock-patch-autospec"
+    priority = -1
+    msgs = {
+        "W1280": (
+            "Uses mock.patch without autospec=True.",
+            name,
+            (
+                "All mock patches should use autospec=True to avoid allowing calls "
+                "to functions that don't exist in production code"
+            ),
+        ),
+    }
+
+    def visit_call(self, call: astroid.node_classes.Call) -> None:
+        if not isinstance(call.func, astroid.node_classes.Attribute):
+            # Not a method call, not applicable
+            return
+
+        if call.func.attrname != "patch":
+            # Some other method, not applicable
+            return
+
+        if not isinstance(call.func.expr, astroid.node_classes.Name):
+            # A method call on a complex expression, literal, or something
+            return
+
+        if call.func.expr.name not in ("mock", "mocker"):
```

Comment:
`class_mocker`, `module_mocker`, `package_mocker` and `session_mocker` are also provided by `pytest-mock`

Reply:
does astroid provide the fully qualified name of object? could we look for `unittest.mock.patch` and [`pytest_mock.plugin.MockFixture._Patcher`](https://github.com/pytest-dev/pytest-mock/blob/master/src/pytest_mock/plugin.py#L144)?

Reply:
Hmm, their docs are a bit sparse but it seems some objects have a `.qname()` function which provides something like that... I will see if that gets us to something more precise like what you are saying.

I guess if that worked, I could check if it's either a call to `_Patcher.__call__` or any normal method of `_Patcher` to cover all the other cases...

That said, I'm not sure if astroid has enough information to do this kind of thing – it only has an AST, so a lot of the names are probably not even resolved to objects yet. I'll see what's available